### DISCUSSION TITLE
How to add indicator to alternate tabbed style?

### DISCUSSION THREAD

**Discussion by MichaIng** [squidfunk/mkdocs-material#3091] [Q&A] [answered]
PyMdown Tabbed added an alternate tab style with a scrollable label bar, which is nice for mobile devices. It is supported by MkDocs material and the old style will be deprecated with next release: https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#tabbed

The new style supports visual indicators via JavaScript: https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/#alternate-style
Are those supposed to be present OOTB when using this alternate style in MkDocs material? I'm currently testing it and don't see any indicator and not a single reference of the related `tabbed-scroll-left`/`tabbed-scroll-right` classes in the generated assets. Else, how can it be enabled? Manually adding the CSS and JS code from PyMdown docs as custom code should work, but I just wanted to assure that this is the intended way. If so probably it can be merged into MkDocs material to be added automatically, or via `mkdocs.yml` setting?

**Comment by facelessuser** [comment, marked as answer]
As far as I am aware, Material does not yet provide JS indicators OOTB. Though, it should be easy enough to pull it from the Pymdown Extensions examples in its documentation.

**Reply by squidfunk** [reply]
They will be added soon.

**Comment by MichaIng** [comment]
Thanks guys. I implemented it now like that: https://github.com/MichaIng/DietPi-Docs/pull/587/files
- The indicator classes in the example were only updated when hovering the bar, but not when scrolling it. So I trigger it on scroll events. E.g. on mobile devices with touch screen, there is no hovering, so the indicators were completely missing and also due to an unlucky screen size one might not see that further tabs are available, so I wanted have them always visible. This could however be adjusted via a single CSS to only show them when being hovered, focused or active (the letter I hope is possible, meaning when a tab label was just clicked).
- I didn't fully understand the example JS scroll state identification, e.g. isn't this:
    ```js
    if (e.type === "mouseover") {
      const scrollWidth = target.scrollWidth - target.clientWidth
      let hscroll = target.scrollLeft
      if (!hscroll) {
        target.scrollLeft = 1
        hscroll = target.scrollLeft
        target.scrollLeft = 0
        if (hscroll) {
          target.classList.add('tabbed-scroll-right')
        }
      } else if (hscroll !== scrollWidth){
        target.classList.add('tabbed-scroll-left', 'tabbed-scroll-right')
      } else if (hscroll) {
        target.classList.add('tabbed-scroll-left')
      }
    }
    ```
    actually this?
    ```js
    if (e.type === "mouseover") {
      const scrollWidth = target.scrollWidth - target.clientWidth
      if (!target.scrollLeft) {
        target.classList.add('tabbed-scroll-right')
      } else if (target.scrollLeft !== scrollWidth){
        target.classList.add('tabbed-scroll-left', 'tabbed-scroll-right')
      } else {
        target.classList.add('tabbed-scroll-left')
      }
    }
    ```
    Or is `target.scrollLeft` somehow forcefully kept `0` or empty when there is no overflow so that this is a hack to check whether it can be scrolled at all? Wouldn't be `((target.scrollWidth - target.clientWidth) > 0)` the cleaner check to know whether to show indicators at all or not?
    However, I simplified it to add/remove the classes individually in our implementation.
- What I still don't like is that it cannot be scrollen via mouse drag. This is expected when no scroll bar is shown, but it is bad when because of unluck the right tab ends exactly at the screen end so that the next tab cannot be clicked, which makes scrolling impossible then. Possibly it makes sense to turn the scroll indicators into actual buttons to scroll left/right, respectively switch the active tab to one left/right?

I hope this helps to further polish things.

**Reply by facelessuser** [reply]
> this is a hack to check whether it can be scrolled at all
 
yes

> So I trigger it on scroll events.

I'm certain it can be improved, I just threw something together to get the feature out the door. The example is not meant to be "the" way to do it, but to give a starting point.

> on mobile devices with touch screen, there is no hovering,

When you click a tab I think they appear, at least in my testing. See the second point above 🙂.

> What I still don't like is that it cannot be scrollen via mouse drag. 

Sure. It can be scrolled on drag on a mobile device, but yeah, not so much with a mouse, but with some JS it can be.

The goal was to keep things looking good on tab overflow, but have it still workable if JavaScript doesn't load, not perfect, but still work reasonably well. Most implementations like this just completely break without JS. JS afterwards is meant to enhance the experience.

I do some web stuff here and there, but I am not a web designer by trade. I was leaving a better implementation up to @squidfunk. I was more providing the Python portion to generate the necessary HTML, provide some kind of workable JS to give it an okay experience for people to run with, but was waiting for an even better solution from @squidfunk.

**Reply by MichaIng** [reply]
No offence, btw, I just share my experience with the example to see whether it can be improved and to potentially support @squidfunk to find a good one for the mkdocs-material 🙂.

> but have it still workable if JavaScript doesn't load

If I'm not mistaken, since scrolling is in the example only done by `scrollIntoView` when hitting the left or right end label, it is not possible without JavaScript on non-touch devices, at least not without a scroll bar 🤔. Another issue I just recognised with `scrollIntoView` is that it sometimes (when the screen width is only slightly smaller than scroll width) scrolls the whole viewport instead of the labels bar, sometimes only that, sometimes both. Similarly the `block` argument, which shouldn't be required as it is for vertical scrolling, applies to the whole viewport when not set to `nearest`. Not sure if there is a way to assure that scrolling is limited to the labels vertical overflow scrolling 🤔.

However, my question is answered, so @squidfunk please ping me when you're start working on the indicators implementation and I'll be happy to test or contribute.

**Reply by facelessuser** [reply]
> If I'm not mistaken, since scrolling is in the example only done by scrollIntoView when hitting the left or right end label, it is not possible without JavaScript on non-touch devices

Scroll wheel on mouse works. Shift + scroll (horizontal).

**Reply by facelessuser** [reply]
On a touchpad, I just two-finger swipe right/left.

**Reply by MichaIng** [reply]
> Scroll wheel on mouse works. Shift + scroll (horizontal).

Nice, I didn't know about this possibility. However, not everyone has a mouse wheel and probably there are others like me who don't know about horizontal mouse wheel scrolling via Shift key 😄. For a bullet prove implementation either a slim scroll bar or buttons are required, IMHO. The letter does not work without JavaScript, so the scroll bar could then be hidden + buttons added via JavaScript, or so.

**Reply by facelessuser** [reply]
I agree all of these things can be solved by JS, but, in the end, the solution is workable if JS fails, and that was the point. I am pretty confident something can be built up with JS to give an even better user experience. I'm not defending things like assuming people should know about horizontal scrolling, I'm only illustrating that things are "functional" even if JS fails, even if it isn't intuitive. Any mouse worth anything should be able to provide horizontal scrolling, but yes, I will 100% admit that not everyone knows how 🙂.

That is also why the feature is marked experimental. We wanted to play with the feature and try to build up, at the very least, a decent experience, before making it official, but to do that, I had to get something out the door that Material could work with.

With all of that said, your feedback has been great, and I am sure it will help the final JS portion of the solution be even better. Though, it is possible that the tab JS solution may be iterative -- I'm not sure what squidfunk has planned yet. Step one is at least to get OOTB functionality that is decent. Step 3, 4, etc. may be to make it, better,  great, etc. 🙂

**Reply by facelessuser** [reply]
@MichaIng If you have some good suggestions, if you can maybe do a mockup, like this: https://codepen.io/facelessuser/pen/VwWdBQX, I'd love to evaluate it as a good suggestion for people in the docs.

**Reply by facelessuser** [reply]
Okay, this is an improvement on what I already had. I added your scroll check fix, but I also made the indicators act like buttons to advance to the next or previous tab.

![tab_button](https://user-images.githubusercontent.com/1055125/136643543-95d44e42-a13d-4030-a800-9869c15e7ff4.gif)

https://codepen.io/facelessuser/pen/VwWdBQX

As a decent demo solution, I think it is pretty decent. You get visual indicators, and they will advance you. Easy to scroll on mobile as swiping works just fine. If the user happens to know about the horizontal scroll, that is yet another way to scroll. Also, when you select the tab, it slides into view. I couldn't ask for much more, but let me know if you come up with some other ideas.

**Reply by facelessuser** [reply]
I kept thinking about this, and so, I've made the codepen even better.

- Now the indicators show on "mouseenter" and "scroll"  of the tab input container.
- When clicking the indicators (which now acts as buttons) the next or pervious (as appropriate) input will be selected scrolling it into view. If there is no next or previous (usually when dealing with the last or first tab respectively) then the current tab will be scrolled into view.
- Buttons seem to also work in touch devices with my limited testing, but even if it doesn't work on some devices, it is already so easy to scroll to new tab on mobile.

**Reply by MichaIng** [reply]
Awesome, yes that is how I would expect it. I was also thinking that those indicators, even if only intended as indicators, may be expected to be buttons by users. When hitting them without being buttons, in cases one hits only the label that is already in view and nothing changes, which does not feel intuitive. Being a button this solves potential issues or non-scrollability without a mouse wheel and feels more intuitive as well.

Whether indicator/buttons shall be visible without hovering, is more personal taste, but e.g. on mobile devices with an unlucky screen width, one might simply not see that there are further tabs, where in some cases you want to assure this, e.g. if a tab contains an important hint or key word. But it is trivial to comment/remove the events from your example code, depending on personal teste/needs.

Btw, for non-JS I personally prefer a styled scroll bar, like when adding this to your mockup:
```css
.tabbed-labels {
  scrollbar-width: thin;
  -ms-overflow-style: unset;
}
.tabbed-labels::-webkit-scrollbar {
  display: unset;
  height: 8px;
}
.tabbed-labels::-webkit-scrollbar-track {
  background: #ccc;
  border-radius: 4px;
}
.tabbed-labels::-webkit-scrollbar-thumb {
  background: #888;
  border-radius: 4px;
}
.tabbed-labels::-webkit-scrollbar-thumb:hover {
  background: #666;
}
.tabbed-labels::-webkit-scrollbar-thumb:active {
  background: #444;
}
```
- On older Firefox, `scrollbar-width` defines the thickness, hence also the height of a horizontal scroll bar.
- With `-webkit-scrollbar`, width and height really mean width and height and affect the thickness only on the matching scroll bar orientation.

It assures scrollability and serves as indicator. Nearly all modern browsers support `-webkit-scrollbar`: https://caniuse.com/css-scrollbar
Could be theoretically overwritten via JavaScript to have the indicator buttons replacing the scrollbar, but for the example in your docs it may be too much code, confusing more than helping as a "code to start with" 😉.

**Reply by facelessuser** [reply]
>  I was also thinking that those indicators, even if only intended as indicators, may be expected to be buttons by users.

They are working as buttons now in my CodePen example. I mentioned that, but maybe it got missed.

> Whether indicator/buttons shall be visible without hovering, is more personal taste

On mobile, personally, I find it sufficient that when you select a tab that it shows there are more with indicators. I can see though, how some may prefer always knowing.

There are also probably ways you could initially show indicators based on media and then monitor media changes to detect mobile and show indicators and such using something maybe like [matchMedia](https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia) or something else. Then you could always show indicators if desired. Personally, I'll leave that complexity up to the user if they want something more dynamic like that.

Scrollbars can also be added with some simple CSS if people just really prefer that.

I do appreciate this discussion though. It kind of made me look closer at things, and honestly, I'm even happier with the new solution. I did make one more update as I forgot to check `((target.scrollWidth - target.clientWidth) > 0)` to kick out of showing indicators.

I'm working to update the official docs with the new implementation examples. As always, it may require tweaking for certain people, but I think it is a general good starting point. For instance, we detect a click on the button using a 15px window. If people had bigger `::before` and `::after` elements, maybe they'd want a window that is larger, but they can play around with it.

**Reply by facelessuser** [reply]
The last one, I swear. Updated solution with a `ResizeObserver`. Now, on page load, if tabs overflow, they will show indicators **and** any time the labels container changes size, indicators will be updated. Basically, anytime tabs overflow, they always show indicators. Scrollbars are not needed at all as there is always an indicator. I don't think IE supports `ResizeObserver`, but I don't care about IE 🤷🏻 .

**Reply by MichaIng** [reply]
> I mentioned that, but maybe it got missed.

No all good, I understood and tested it and just wanted to underline that this satisfies my thoughts 😉.

> show indicators based on media

This is really up to the use case then, how much tabs are used, how long the labels, how they are arranged etc, so perfectly fine to have a default which includes the hover effect and skips media detection. On larger screens the labels fit onto the screen more often, so having them always visible has less often an impact, so if indicators shall only be always visible on mobile phones/touch screens and some exceptions (small non-touch screens, large tab bars) are acceptable, media detection could be skipped as well and only the mouse events need to be removed from the example, which should be trivial.

> Scrollbars can also be added with some simple CSS if people just really prefer that.

Yes, the idea was to show it by default (only CSS loaded) and hide it via JavaScript, so that on CSS-only clients one has the scroll bar and on JavaScript-capable clients one has the scroll buttons and the scroll bar hidden. If the scroll bar is generally sufficient, of course the whole JavaScript can be skipped.

> For instance, we detect a click on the button using a 15px window.

That is perfectly fine to have some example button CSS, which can then be tweaked to personal needs, e.g. also the shape or even have an image/SVG for those 🙂.
_________
#### EDIT
> Updated solution with a `ResizeObserver`.

Ah nice, this was also one minor glitch I found before. Often, when resizing the window, a scroll event is included, but not always. Especially when the buttons are shown on hover only, it is a pretty rare cases that the window is resized while the bar is hovered, but still good to have the case covered, especially when buttons are always shown.

**Reply by facelessuser** [reply]
Pymdown Extensions documentation has implemented the new solution in our own docs and has updated demo code with the new solution as well.

I like the fact that indicator buttons no longer require a hover. They just auto show on page load and update on resize and scroll. I'm also really liking that the indicators function as buttons. I think this all makes the alternate styling not just usable, but a pretty decent solution.

Thanks again, and with this, I'm sure we've made @squidfunk's job easier as we've given him a number of directions he can go in now.

**Reply by MichaIng** [reply]
Since mkdocs-material enforces the new style now without having indicators implemented, I went on and updated our indicator/button implementation according to the pymdown docs, many thanks for providing those 👍. Some remarks:
- `selected.click()`, i.e. when the currently selected tab is the first or last while clicking the left resp. right button, has no effect. I guess it was supposed to scroll it into view in case it is out of view, which intuitively is what I'd expect. But that can be done much easier than triggering an event: `target.scrollLeft = target.scrollWidth;` resp. `target.scrollLeft = 0;`
- The detection whether the button has been clicked or not is a bit fragile: Whether the first or the last 15 pixels or the labels bar matches the buttons or not depends on quite same other style factors. E.g. mkdocs-material currently allows the labels bar to overflow the parent tabs div and container element on small screens, so the last 15 pixels are outside of the container behind the button. Also it strangely adds a right padding to the `::after` element, not sure what this is for when no such buttons/indicators have been implemented yet, so the button visually further moves to the left. The overflow and padding additionally has the bad visual effect that labels are partly visible left to the left button and right to right button, instead of nicely ending in the alpha gradient of the buttons, as intended by the example CSS.
  - Not sure whether there is a way to cheaply get the size of the button instead of using a fixed pixel number? There are reasonably consequently `em` units used for font and related element sizes, but I guess those cannot be used in JavaScript 🤔. I was already thinking whether it would be more robust to have "static" scroll buttons on every tabs bar to explicitly add click event listeners to those (instead of the parent tabs bar) and then only show/hide them based on scroll state.
- The `ResizeObserver` is required to even get the resize event? Wouldn't it be sufficient to check for "global" window resizing only instead of adding this to every tabs bar? Or was this intended to be failsafe when other style changes lead to the resizing of individual tab bars?
- Btw, the `if (!e.target.matches('.tabbed-labels')) {` in `checkScroll` is just to be failsafe is something else is calling this function, right? Or do e.g. child elements or so as well unnecessarily trigger the event?

<details>
<summary>For reference, my JavaScript now looks like this:</summary>

```js
const tabOverflow = () => {
  // Identify whether tab bar can be scrolled left or right and apply button classes
  const checkScroll = e => {
    const target = e.target;
    if (target.scrollLeft) {
      target.classList.add('tabbed-scroll-left');
    } else {
      target.classList.remove('tabbed-scroll-left');
    }
    if (target.scrollLeft < (target.scrollWidth - target.clientWidth - 3)) {
      target.classList.add('tabbed-scroll-right');
    } else {
      target.classList.remove('tabbed-scroll-right');
    }
  }

  // Check again when tab bar is resized
  const onResize = new ResizeObserver(entries => {
    entries.forEach(entry => {
      checkScroll({target: entry.target, type: 'resize'});
    })
  })

  // Change tab when clicking button
  const tabChange = e => {
    const target = e.target;
    if (target.classList.contains('tabbed-scroll-right') && e.offsetX >= target.offsetWidth - 20) {
      const sib = target.closest('.tabbed-alternate').querySelector('input:checked').nextSibling;
      if (sib && sib.tagName === 'INPUT') {
        sib.click();
      } else {
        target.scrollLeft = target.scrollWidth;
      }
    } else if (target.classList.contains('tabbed-scroll-left') && e.offsetX <= 20) {
      const sib = target.closest('.tabbed-alternate').querySelector('input:checked').previousSibling;
      if (sib && sib.tagName === 'INPUT') {
        sib.click();
      } else {
        target.scrollLeft = 0;
      }
    }
  }

  const labels = document.querySelectorAll('.tabbed-alternate > .tabbed-labels');
  labels.forEach(el => {
    checkScroll({target: el, type: 'resize'});
    onResize.observe(el);
    el.addEventListener('resize', checkScroll);
    el.addEventListener('scroll', checkScroll);
    el.addEventListener('click', tabChange);
  })
}

// Smooth scroll tab into view when changed
const tabScroll = () => {
  const tabs = document.querySelectorAll('.tabbed-alternate > input');
  for (const tab of tabs) {
    tab.addEventListener('change', () => {
      const label = document.querySelector(`label[for=${tab.id}]`);
      label.scrollIntoView({block: 'nearest', inline: 'nearest', behavior: 'smooth'});
    })
  }
}

// Register events
tabOverflow();
tabScroll();
```
</details>
<details>
<summary>And CSS to resolve/overwrite the conflicting `mkdocs-material` styles:</summary>

```css
@media screen and (max-width: 44.9375em) {
    .tabbed-alternate > .tabbed-labels {
        max-width: 100% !important;
        margin: 0 !important;
        padding: 0 !important;
    }
}

.tabbed-alternate > .tabbed-labels.tabbed-scroll-left::before {
    display: inline-block;
    font-size: 0.9em;
    position: absolute;
    top: 0.75em;
    left: 0;
    padding-left: 0;
    padding-right: 0.64em;
    color: gray;
    background: linear-gradient(to right, rgb(255, 255, 255) 75%, rgba(255, 255, 255, 0));
    content: "\25C0";
    cursor: pointer;
}

.tabbed-alternate > .tabbed-labels.tabbed-scroll-right::after {
    display: inline-block;
    font-size: 0.9em;
    position: absolute;
    top: 0.75em;
    right: 0;
    padding-left: 0.64em;
    padding-right: 0;
    color: gray;
    background: linear-gradient(to right, rgba(255, 255, 255, 0), rgb(255, 255, 255) 25%);
    content: "\25B6";
    cursor: pointer;
}
```
</details>

**Reply by squidfunk** [reply]
> Also it strangely adds a right padding to the ::after element, not sure what this is for when no such buttons/indicators have been implemented yet, so the button visually further moves to the left.

It's a hack. The code is in fact very thoroughly commented 😉

https://github.com/squidfunk/mkdocs-material/blob/cc04ddead0622ca9ccfe2bb06fc51ecacc2173a5/src/assets/stylesheets/main/extensions/pymdownx/_tabbed.scss#L199-L202

**Reply by MichaIng** [reply]
Ah thanks for explanation. This conflicts with the scroll buttons/indicators then, the way they are implemented in the pymdown example. This would be an argument for the more explicit implementation as two always existing button elements with dedicated click event handlers, being shown/hidden based on scroll state.

**Reply by facelessuser** [reply]
1. Keep in mind that `tabbed_alternate` is still experimental, even if Material has decided to go full in on this.
2. Material is going for a style similar to tabs on GitHub's site, where no tab indicators are used. I think this works pretty well except when you nest them within Admonitions and such.
3. The "suggested" starting style via pymdown-extensions (which uses indicators) is still a work in progress as the feature is still experimental.

I know I've tweaked the style even and code even from when I initially provided the example in the documentation.

With all of that said, I'm definitely willing to have a conversation on how to make the "suggested" style better, but I'll need some minimal working examples illustrating what is meant. You can help by maybe providing a CodePen? I have a  base one here: https://codepen.io/facelessuser/pen/VwWdBQX.

I believe you made one point about the indicator not scrolling the far left or far right tab into view if the tab is already selected. This is known and currently, an accepted behavior at this point. I'm open to suggestions if people want to make it better, but the example right now was "good enough" for my use. Keep in mind this is a style to get people "started", but not necessarily the "best".

**Reply by MichaIng** [reply]
> Material is going for a style similar to tabs on GitHub's site, where no tab indicators are used.

Do you have a link how it looks like? On the GitHub website I cannot find any similar tabs bars, the top navigation get's `...` to expand a menu with additional links and GitHub's Markdown does not support `tabbed`. Or do you mean GitHub pages?

> I think this works pretty well except when you nest them within Admonitions and such.

Currently it is very fragile, users can miss at all that there are more tabs available, auto-scrolling is done unpredictable depending much on the browser etc. But I mentioned this above already. If it was wanted without scroll bar or buttons, then IMHO it would need to be assured that previous/next labels are always at least partly visible and clickable to the left/right of the bar.

> You can help by maybe providing a CodePen? I have a base one here:

Open your example, scroll via shift + wheel to the right, then click the left indicator button: Nothing happens, at least not on Chromium based browsers. Then replace `updated = selected` with `target.scrollLeft = target.scrollWidth` resp. `target.scrollLeft = 0` (and skip the `click()`) so that the click will scroll you to the right resp. left end to get the selected tab fully in view. This is all that I found as issue with your example code. The other points are more questions or related to the mkdocs-material styles, not present with the example style 🙂.

Btw, my words may read a bit harsh: Please do not take them like that or as offence. I'm happy to see development on your projects, which I am able to freely use, and if I wouldn't like any of it, I was free to stick with older versions or use other solutions. So take this as hopefully constructive criticism/feedback from one user, to polish the new tabs style implementation. I guess now with this being standard in `mkdocs-material` there will be more feedback, so we can see whether I am picky or whether others see issues similarly, and maybe there are even better ideas to address the issues or enhance user experience 🙂.

**Reply by facelessuser** [reply]
> Open your example, scroll via shift + wheel to the right, then click the left indicator button: Nothing happens, at least not on Chromium based browsers.

Yes, if the tab is already selected, it doesn't currently scroll into view which is known. With that said, I will definitely take a look at your suggestion as this is an area that could use improvement, it's just been low on my list right now.

> Btw, my words may read a bit harsh: Please do not take them like that or as offence. 

I didn't really take offense, I kind of read over it quickly and just fired off a general response simply to clarify it's known the solution may not be perfect and I still consider the implementation as experimental for this reason. But, I'm more than happy to collaborate to make it better, so thank you for sharing your thoughts, and I'll definitely look into this when I get a moment 🙂.

**Reply by MichaIng** [reply]
> Yes, if the tab is already selected, it doesn't currently scroll into view which is known.

Ah, I though that the `selected.click()` (actually `updated = selected` + `updated.click()`) was supposed to scroll it into view then, as this code is only called that way when there is no previous/next tab to select, and as of the code comment `// Make sure the current, selected input is scrolled into view.`. Then I wonder what `selected.click()` was supposed to do? However, yes consider to do a full left/right scroll in this case, so that users can directly see that the most left/right tab is selected already. But it is a rare care anyway, I mean the selected tab can only be out of view when it was manually scrolled via touch drag or shift + mouse wheel already.

> I didn't really take offense

Good, sometimes the text lacks to transport the mood/how it was meant 🙂.

**Reply by squidfunk** [reply]
Yes, the implementation still needs some polishing, but it's already way better than before, so please consider the new implementation the polishing of the old one. This project is so huge, when you finished polishing everything, you can make another pass and another pass etc. I'm happy to collect some feedback on the tabs implementation, but currently, it's not really a priority. IMHO the implementation is already pretty good, it works without JavaScript, on mobile, and with keyboard navigation, thus it's also accessible, which is also a priority.

You can create a new discussion to collect some feedback, but unless something doesn't work, I'm currently not considering this to be an issue/bug/problem. As you already said, you're free to use an older version of Material for MkDocs, or port the old styles to some extra CSS. Material for MkDocs always tries to provide reasonable defaults, but bringing your own is easily possible and encouraged.

**Reply by facelessuser** [reply]
> But it is a rare care anyway, I mean the selected tab can only be out of view when it was manually scrolled via touch drag or shift + mouse wheel already.

Yep, that's exactly why I didn't worry about it much. But, I agree that it could be better. I've kind of been iterating on it little by little. Overall, I think it works better than the old way, but yeah more polish is still probably needed.

**Reply by MichaIng** [reply]
Okay thanks for clarifying. Probably it's really me being picky about that, or our documentation where it is essential that visitors see all available tabs, respectively an indication that there are more tabs.

Defending the old way: At least every tab was always visible and always selectable with a single click/tap. Whether having those across multiple lines or not is a purely visual aspect that one can like or not. At least me and one team member weren't unhappy with those being shown across multiple lines, i.e. we never saw this as a issue which needs to be addressed. But this is really a question of personal taste, and then of course whether the implications of a single line solution are seen as as larger downside than multi-line labels were. But I'm closing the feedback on this now 😄. This decision has been made and we are always free to implement it differently if we wanted.

**Reply by facelessuser** [reply]
I will say, the Tabbed extension has not necessarily abandoned the old way. That is still too be decided, which is separate from whether Material provides default styling for it

**Reply by squidfunk** [reply]
Material for MkDocs never supported all Python Markdown Extensions. We support a large subset of them, but for some not the entirety of available options, like for example line anchors etc. However, @facelessuser does a very good job explaining how you can add them to your documentation. We might add support for more features gradually, but we will probably never support all of them.

This is an opinionated project. I understand that it generates a website that you want to give to your users, and when that website doesn't match 100% of your expectations, you want to change things because it represents your brand. However, so do all other users. I'm trying to provide sensible defaults, but it's absolutely impossible to make everybody happy.

Recommended read:

- #2399
